### DISCUSSION TITLE
How do I scope to an enum and then do .ifCaseLet?

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-composable-architecture#1420] [Q&A] [answered]
I'm working with the protocol beta branch.

I had a feature like this:

```
public enum State: Equatable {
  case login(LoginFeature.State)
  case mainApp(MainAppFeature.State)
  ...
}
```

In my reducer I was able to switch between states and then use `.ifCaseLet` to scope to the various cases...

```
public var body: some ReducerProtocol<State, Action> {
  Reduce { state, action in
    case .something:
      state = .login(.init())
      return .none
  }
  .ifCaseLet(/State.login, action: /Action.login) {
    LoginFeature()
  }
  .ifCaseLet...
}
```

But now I wanted to change it so that the `State` is now a struct that contains this enum...

```
public struct State: Equatable {
  enum AppMode: Equatable {
    case login(LoginFeature.State)
    case mainApp(MainAppFeature.State)
    ...
  }

  var appMode: AppMode
}
```

But I'm struggling with how to update the reducer to work with this.

I tried creating a `Scope` reducer and then doing `.ifCaseLet` on that but I think I got that wrong. 😅 

The compiler complains it cannot type-check the reducer in a reasonable time.

In the old way I think I would have done a scope and then case on the Login reducer but now I'm not so sure with this.

Any guidance would be appreciated. Thanks

**Comment by mbrandonw** [comment, marked as answer]
This is unfortunately a blind spot in the library right now. We don't provide an ergonomic tool to first scope to an enum field and then further scope to a case. There is  a [workaround](https://github.com/pointfreeco/swift-composable-architecture/discussions/1296#discussioncomment-3487107) though, which you can use until we come up with a better tool.

**Reply by mbrandonw** [reply]
And you should also read Stephen's follow comment (and link) in that thread.

**Reply by oliverfoggin** [reply]
Thanks, I'll have a read now.

**Reply by oliverfoggin** [reply]
Oh man! I was so close. I just didn't get the order of Scope and EmptyReducer and ifCaseLet correct. 😄
